File: C_plus_plus/Sum_Root_to_Leaf_Numbers.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sum;
    int sumNumbers(TreeNode *root) {
        sum = 0;
        if(root == NULL) return 0;
        travel(root,0);
        return sum;
    }
    void travel(TreeNode *node,int pre){
        if(node->left == NULL && node->right == NULL)
            sum += pre*10+node->val;
        else{
            if(node->left != NULL) travel(node->left,pre*10+node->val);
            if(node->right != NULL) travel(node->right,pre*10+node->val);
        }
    }
};

File: java/Sum_Root_to_Leaf_Numbers.java
// DFS method
// Time is O(n), space is O(n).

public class Solution {
    public int sumNumbers(TreeNode root) {
        return dfs(root,0);
    }
    public int dfs(TreeNode root, int preValue) {
        if (root == null) return 0;
        if (root.left == null && root.right == null) return preValue*10 + root.val;
        return dfs(root.left, preValue*10 + root.val) + dfs(root.right, preValue*10 + root.val);
        
    }
}
